### PR TITLE
Seperate trie commit and hash

### PR DESCRIPTION
**Pull request by hqjang-pepper**
## Proposed changes
Seperate Trie committer and hasher
- related ethereum PRs
  - https://github.com/ethereum/go-ethereum/pull/20481
  - https://github.com/ethereum/go-ethereum/pull/21351

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hqjang-pepper** [comment]
### Benchmark Hash and commit



- #### Hash

```
Before
BenchmarkHashFixedSize
BenchmarkHashFixedSize/10
BenchmarkHashFixedSize/10-10 	           27526	     43064 ns/op	   14005 B/op	     157 allocs/op
BenchmarkHashFixedSize/100
BenchmarkHashFixedSize/100-10         	   10000	    101392 ns/op	   65680 B/op	     777 allocs/op
BenchmarkHashFixedSize/1K
BenchmarkHashFixedSize/1K-10          	    1190	   1004009 ns/op	  657317 B/op	    7782 allocs/op
BenchmarkHashFixedSize/10K
BenchmarkHashFixedSize/10K-10         	     274	   4429889 ns/op	 6905728 B/op	   79660 allocs/op
BenchmarkHashFixedSize/100K
BenchmarkHashFixedSize/100K-10        	      27	  38932048 ns/op	68239165 B/op	  794500 allocs/op

After
BenchmarkHashFixedSize
BenchmarkHashFixedSize/10
BenchmarkHashFixedSize/10-10 	           37155	     32239 ns/op	   11942 B/op	     135 allocs/op
BenchmarkHashFixedSize/100
BenchmarkHashFixedSize/100-10         	    7766	    153061 ns/op	   57988 B/op	     671 allocs/op
BenchmarkHashFixedSize/1K
BenchmarkHashFixedSize/1K-10          	     798	   1505582 ns/op	  584060 B/op	    6728 allocs/op
BenchmarkHashFixedSize/10K
BenchmarkHashFixedSize/10K-10         	      66	  18371760 ns/op	 6224453 B/op	   69396 allocs/op
BenchmarkHashFixedSize/100K
BenchmarkHashFixedSize/100K-10        	       6	 177527841 ns/op	61796261 B/op	  693042 allocs/op
```


- #### Commit
```
Before
BenchmarkCommitAfterHashFixedSize
BenchmarkCommitAfterHashFixedSize/10
BenchmarkCommitAfterHashFixedSize/10-10 	           30988	     38790 ns/op	   36083 B/op	     208 allocs/op
BenchmarkCommitAfterHashFixedSize/100
BenchmarkCommitAfterHashFixedSize/100-10         	    8286	    145892 ns/op	  174370 B/op	    1022 allocs/op
BenchmarkCommitAfterHashFixedSize/1K
BenchmarkCommitAfterHashFixedSize/1K-10          	     828	   1485660 ns/op	 1701478 B/op	   10143 allocs/op
BenchmarkCommitAfterHashFixedSize/10K
BenchmarkCommitAfterHashFixedSize/10K-10         	      74	  14636495 ns/op	18696716 B/op	  103985 allocs/op
BenchmarkCommitAfterHashFixedSize/100K
BenchmarkCommitAfterHashFixedSize/100K-10        	       7	 148162095 ns/op	180172019 B/op	 1038887 allocs/op

After
BenchmarkCommitAfterHashFixedSize
BenchmarkCommitAfterHashFixedSize/10
BenchmarkCommitAfterHashFixedSize/10-10 	           60080	     20030 ns/op	   30914 B/op	     189 allocs/op
BenchmarkCommitAfterHashFixedSize/100
BenchmarkCommitAfterHashFixedSize/100-10         	   12500	     96729 ns/op	  151233 B/op	     918 allocs/op
BenchmarkCommitAfterHashFixedSize/1K
BenchmarkCommitAfterHashFixedSize/1K-10          	    1052	   1188825 ns/op	 1473359 B/op	    9108 allocs/op
BenchmarkCommitAfterHashFixedSize/10K
BenchmarkCommitAfterHashFixedSize/10K-10         	      94	  11857340 ns/op	16368251 B/op	   93862 allocs/op
BenchmarkCommitAfterHashFixedSize/100K
BenchmarkCommitAfterHashFixedSize/100K-10        	       9	 121488097 ns/op	157054739 B/op	  937567 allocs/op
```

File: storage/statedb/hasher.go [hunk 1]
```diff
@@ -76,226 +75,126 @@ func returnHasherToPool(h *hasher) {
...
-			if n.Children[i] != nil {
-				numRootChildren++
-				go func(i int, n node) {
-					childHasher := newHasher(h.onleaf)
-					defer returnHasherToPool(childHasher)
-					collapsedFromChild, cachedFromChild := childHasher.hash(n, db, false)
-					hashResultCh <- hashResult{i, collapsedFromChild, cachedFromChild}
-				}(i, n.Children[i])
-			}
-		}
```
Comment (jeongkyun-oh) [inline storage/statedb/hasher.go:226]: This part is optimized in order to calculate root hash more efficiently. Was this considered?

The related PR: https://github.com/klaytn/klaytn/pull/310
cc @kjhman21 @ehnuje

File: storage/statedb/trie.go [hunk 2]
```diff
@@ -78,13 +78,12 @@ func (t *Trie) newFlag() nodeFlag {
 // not exist in the database. Accessing the trie loads nodes from db on demand.
 func NewTrie(root common.Hash, db *Database) (*Trie, error) {
 	if db == nil {
-		panic("statedb.NewTrie called without a database")
+		panic("trie.New called without a database")
```
Comment (jeongkyun-oh) [inline storage/statedb/trie.go:81]: This change is not necessary, right?

File: storage/statedb/trie.go [hunk 3]
```diff
@@ -55,14 +56,13 @@ type LeafCallback func(paths [][]byte, hexpath []byte, leaf []byte, parent commo
 
 // Trie is a Merkle Patricia Trie.
 // The zero value is an empty trie with no database.
-// Use NewTrie to create a trie that sits on top of a database.
+// Use New to create a trie that sits on top of a database.
```
Comment (jeongkyun-oh) [inline storage/statedb/trie.go:59]: This change is not necessary, right?

File: storage/statedb/trie.go [hunk 4]
```diff
@@ -504,11 +503,7 @@ func (t *Trie) resolve(n node, prefix []byte) (node, error) {
 
 func (t *Trie) resolveHash(n hashNode, prefix []byte) (node, error) {
 	hash := common.BytesToHash(n)
-	node, fromDB := t.db.node(hash)
-	if t.prefetching && fromDB {
-		memcacheCleanPrefetchMissMeter.Mark(1)
-	}
```
Comment (jeongkyun-oh) [inline storage/statedb/trie.go:510]: Why this is removed?

File: storage/statedb/trie.go (hunk 2 above)
Comment (hqjang-pepper) [reply storage/statedb/trie.go:81]: Oh you are right. thanks

File: storage/statedb/trie.go (hunk 3 above)
Comment (hqjang-pepper) [reply storage/statedb/trie.go:59]: Oh you are right. thanks

File: storage/statedb/trie.go (hunk 4 above)
Comment (hqjang-pepper) [reply storage/statedb/trie.go:510]: My bad. I'll roll back this change.

File: storage/statedb/committer.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+			case *shortNode:
+				if child, ok := n.Val.(valueNode); ok {
+					c.onleaf(nil, nil, child, hash, 0)
+				}
+			case *fullNode:
+				// For children in range [0, 15], it's impossible
+				// to contain valueNode. Only check the 17th child.
+				if n.Children[16] != nil {
+					c.onleaf(nil, nil, n.Children[16].(valueNode), hash, 0)
+				}
```
[the remark this answers is not in the document]
Comment (hqjang-pepper) [reply storage/statedb/committer.go:241]: Yes I think there's no specific problem in here. 
Same as the existing onleaf method.

**Comment by kjhman21** [comment]
@hqjang-pepper yes.

**Comment by blukat29** [comment]
Won't do in this version. closing.

**Pull request closed without merging**